packages/pyright-internal/src/analyzer/dataClasses.ts

```diff
@@ -654,6 +673,93 @@ export function synthesizeDataClassMethods(
     );
 }
 
+// Validates converter and, if valid, returns its input type. If invalid,
+// fieldType is returned.
+function getConverterInputType(
+    evaluator: TypeEvaluator,
+    converterNode: ArgumentNode,
+    fieldType: Type,
+    fieldName: string
+): Type {
+    const converterType = evaluator.getTypeOfExpression(converterNode.valueExpression).type;
+
+    if (!isFunction(converterType) && !isOverloadedFunction(converterType)) {
+        return fieldType;
+    }
+
+    // Create synthesized function of the form Callable[[T], fieldType] which
+    // will be used to check compatibility of the provided converter.
+    const typeVar = TypeVarType.createInstance('__converterInput');
+    typeVar.scopeId = evaluator.getScopeIdForNode(converterNode);
+    const targetFunction = FunctionType.createSynthesizedInstance('');
+    targetFunction.details.declaredReturnType = fieldType;
```

Comment:
You should set the details.typeVarScopeId for targetFunction as well. It should share the scope ID with its TypeVar.